File: Assets/Resources/Skill/FireBallCaster.js
﻿#pragma strict
import PhotonNetwork;
import NetworkManager;
var skillIndex: int = 0;
var skillName: String = 'Fire Ball';
var chantTime: double = 1000;
var isChanting: boolean = false;
var owner: GameObject;
private var epochStart: System.DateTime = new System.DateTime(1970, 1, 1, 0, 0, 0, System.DateTimeKind.Utc);
private var eventCenter: EventCenter;
private var coolDownTimer: CoolDownTimer;
private var isButtonPressed: boolean = false;
private var timeStartChanting: double = 0;

function Awake () {
	coolDownTimer = GetComponent(CoolDownTimer);
	eventCenter = GameObject.FindWithTag('EventCenter').GetComponent(EventCenter);
	eventCenter.RegisterListener(eventCenter, 'skillButtonDown', gameObject, OnSkillButtonDown);
	eventCenter.RegisterListener(eventCenter, 'skillButtonUp', gameObject, OnSkillButtonUp);
	eventCenter.RegisterListener(eventCenter, 'leftButtonPressed', gameObject, OnPlayerMove);
	eventCenter.RegisterListener(eventCenter, 'rightButtonPressed', gameObject, OnPlayerMove);
}
function Start() {
    
}
function Update () {
    // FIXME: this assignment should be removed.
    owner = GameObject.FindWithTag('Player');

	if(false == isButtonPressed) {	
		isChanting = false;
		return;
	}
	var timestamp: double = (System.DateTime.UtcNow - epochStart).TotalMilliseconds;
	if(false == isChanting) {
		if(true == coolDownTimer.GetIsCoolDownFinished()) {
			isChanting = true;
			var startData = new ChantingEventData('start', owner, gameObject);
			eventCenter.CastEvent(eventCenter, 'startChanting', startData);
			timeStartChanting = timestamp;
		}
	} else {
		if((timeStartChanting + chantTime) < timestamp) {
			coolDownTimer.StartCoolDown();
			isChanting = false;
			var stopData = new ChantingEventData('stop', owner, gameObject);
			eventCenter.CastEvent(eventCenter, 'stopChanting', stopData);
			Cast();
		}
	}
}
var OnSkillButtonDown = function(e: SbiEvent) {
	var index: int = System.Convert.ToInt32(e.data);
	if(skillIndex != index) return;
	isButtonPressed = true;
};
var OnSkillButtonUp = function(e: SbiEvent) {
	var index: int = System.Convert.ToInt32(e.data);
	if(skillIndex != index) return;
	isButtonPressed = false;
    // FIXME: this Cast should be removed.
	Cast();
};
var OnPlayerMove = function(e: SbiEvent) {
	isChanting = false;
	isButtonPressed = false;
};
private function Cast() {
    NetworkManager.Instance.Instantiate(
		'Skill/FireBallBullet', owner.transform.position, owner.transform.rotation , 0
	);
}

File: Assets/Script/Skill/FireBallCaster.cs
using UnityEngine;

public class FireBallCaster : MonoBehaviour {
	public int skillIndex = 0;
	public string skillName = "Fire Ball";
	public GameObject owner;
	public GameObject Owner {
		get {
			if((null == owner) && (null != transform.parent) && (null != transform.parent.parent)) {
				owner = transform.parent.parent.gameObject;
			}
			return owner;
		}
		set { owner = value; }
	}

	private SkillProperties skillProperties;
	private CoolDownTimer coolDownTimer;
	private ChantTimer chantTimer;
	private bool isButtonPressed = false;

	void Awake () {
		skillProperties = GetComponent<SkillProperties>();
		coolDownTimer = GetComponent<CoolDownTimer>();
		chantTimer = GetComponent<ChantTimer>();
		EventManager.Instance.RegisterListener(EventManager.Instance, "skillButtonDown", gameObject, OnSkillButtonDown);
		EventManager.Instance.RegisterListener(EventManager.Instance, "skillButtonUp", gameObject, OnSkillButtonUp);
		EventManager.Instance.RegisterListener(EventManager.Instance, "leftButtonPressed", gameObject, OnPlayerMove);
		EventManager.Instance.RegisterListener(EventManager.Instance, "rightButtonPressed", gameObject, OnPlayerMove);
		EventManager.Instance.RegisterListener(EventManager.Instance, "casting", gameObject, OnCasting);
	}
	
	void Update () {
		if(false == isButtonPressed) {	
			if(true == chantTimer.isChanting) { chantTimer.StopChanting(); }
			return;
		}
		if(false == chantTimer.isChanting) {
			if(true == coolDownTimer.GetIsCoolDownFinished()) {
				chantTimer.StartChanting();
			}
		} else {
			if(true == chantTimer.GetIsChantingFinished()) {
				coolDownTimer.StartCoolDown();
				chantTimer.StopChanting();
				CastRPC();
			}
		}
	}
	private void OnSkillButtonDown(SbiEvent e) {
		if(GameManager.Instance.GetPlayerCharacter() != Owner) { return; }
		SkillButtonEventData data = e.data as SkillButtonEventData;
		if(skillIndex != data.index) {
			chantTimer.StopChanting();
			return;
		}
		isButtonPressed = true;
	}
	private void OnSkillButtonUp(SbiEvent e) {
		if(GameManager.Instance.GetPlayerCharacter() != Owner) { return; }
		SkillButtonEventData data = e.data as SkillButtonEventData;
		if(skillIndex != data.index) return;
		isButtonPressed = false;
	}
	private void OnPlayerMove(SbiEvent e) {
		if(GameManager.Instance.GetPlayerCharacter() != Owner) { return; }
		chantTimer.StopChanting();
		isButtonPressed = false;
	}
	private void CastRPC() {
		if(null == gameObject.transform.parent) { return; }
		SkillHandler handler = gameObject.transform.parent.gameObject.GetComponent<SkillHandler>();
		handler.CastRPC(skillProperties.skillId);
	}
	private void OnCasting(SbiEvent e) {
		CastingEventData data = (CastingEventData)e.data;
		if((Owner != data.role) || (skillProperties.skillId != data.skillId)) {
			return;
		}
        GameObject target = Owner.GetComponent<LookAt>().target;
        Quaternion direction = Quaternion.LookRotation(target.transform.position - Owner.transform.position);

		GameObject fireBallBulletGameObject = (GameObject)GameObject.Instantiate(
			Resources.Load("Prefab/Skill/FireBallBullet"), Owner.transform.position, direction
		);

		fireBallBulletGameObject.GetComponent<Faction>().SetFaction(
			Owner.GetComponent<Faction>().GetFaction()
		);
	}
}
